### DISCUSSION TITLE
How to use blocks correctly

### DISCUSSION THREAD

**Discussion by EduardGurman** [Azure/ipam#290] [General]
Hey, 
I'm currently trying to check if the solution answers our needs and while playing with it, I found something strange. 
I've crated spaces for all of our subscriptions and blocks for each Vnet - (each block contains 1 Vnet), then I've made the association between each block and its Vnet.
While looking at the Vnets and subnets tabs, I can see the correct total/allocated ip's, but from the spaces and blocks tabs all the items are fully allocated.
Am I doing something wrong ?

**Comment by DCMattyG** [comment]
Good morning @EduardGurman, I'm happy to give a quick overview of how Spaces/Blocks work, as well as some suggestions of the best way to utilize them.

First and foremost, you should most certainly **not** use one Space per Subscription.

A **Space** is effectively a logical container where you can only have unique, non-overlapping IP space. Some customers may only end up having a single **Space**, and that is perfectly OK. There are many companies that really don't practice IP reuse across multiple environments, but there are some whom use it extensively. For instance, let's say you are a company who has a Staging and Production environment and you like to keep both environment identical to the point that you even use the exact same IP's across both, but that is fine since the environments aren't ever connected to each other and as such there won't be any routing issues because of the overlap...in this example you might have two **Spaces**, one for Staging and the other for Production.

**Blocks** live within **Spaces** and are effectively "superblocks" or large CIDR's which are assigned by the Networking/NetOps team for use in different environments and then can be further carved out into various Virtual Networks. Another example here is let's say you have 10.1.0.0/16 assigned for use in Azure West US and then 10.2.0.0/16 assigned for use in Azure East US, in this situation you would have two **Blocks** within a **Space**, one for each aforementioned CIDR range.

As far as the utilization numbers roll up, this is how that logic works:

1. Space utilization is based on the total CIDR space of all of it's child Blocks, and how much of that CIDR space is allocated to Virtual Networks associated to each of the Blocks.
2. Block utilization is based on it's own CIDR space and how much of that is currently allocated to Virtual Networks associated to the Block.
3. Virtual Network utilization is based on the total size of all IP address spaces assigned to the vNET and how much of that has been allocated to Subnets created within that vNET.
4. Subnet utilization is based on the total size of the Subnet and how many endpoints are associated to that Subnet, less the 5 required IP's from the Azure side.

So, I hope that has helped to clear things up a little for you. As always I'm more than happy to elaborate further on any of the above points if required. Please let me know if you need anything else and we'll be here to support you however we can.

Have a wonderful week!

**Reply by EduardGurman** [reply]
Ohh wow, I was totally off.
Thank you very much for the detailed answer, I'll have a talk with our NetOps team.

**Reply by DCMattyG** [reply]
You bet, so glad that helped!
